File: truevfs-ext/truevfs-ext-pace/src/main/java/net/java/truevfs/ext/pace/PaceManager.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pace;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import javax.annotation.concurrent.ThreadSafe;
import net.java.truecommons.shed.BitField;
import net.java.truecommons.shed.HashMaps;
import net.java.truevfs.comp.jmx.JmxManager;
import net.java.truevfs.kernel.spec.*;
import net.java.truevfs.kernel.spec.sl.FsManagerLocator;

/**
 * The pace manager.
 * 
 * @author Christian Schlichtherle
 */
@ThreadSafe
final class PaceManager extends JmxManager<PaceMediator> {

    private static final int INITIAL_CAPACITY = HashMaps.initialCapacity(
            PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_DEFAULT_VALUE + 1);

    private final Collection<FsController> evicted = new ConcurrentLinkedQueue<>();
    private final MountedControllerMap mounted = new MountedControllerMap(evicted);
    private final SyncedControllerSet synced = new SyncedControllerSet(mounted);

    PaceManager(PaceMediator mediator, FsManager manager) {
        super(mediator, manager);
    }

    int getMaximumFileSystemsMounted() {
        return mounted.max;
    }

    void setMaximumFileSystemsMounted(final int max) {
        if (max < PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_MINIMUM_VALUE)
            throw new IllegalArgumentException();
        mounted.max = max;
    }

    @Override
    protected Object newView() { return new PaceManagerView(this); }

    /**
     * If the number of mounted archive files exceeds
     * {@link #getMaximumFileSystemsMounted()}, then this method
     * {@linkplain #sync syncs} the least recently used (LRU) archive files
     * which exceed this value.
     *
     * @param  controller the controller for the file system to retain mounted
     *         for subsequent access.
     */
    void retain(final FsController controller) throws FsSyncException {
        final Iterator<FsController> i = evicted.iterator();
        if (!i.hasNext()) return;
        final FsManager manager = FsManagerLocator.SINGLETON.get();
        final FsMountPoint mp = controller.getModel().getMountPoint();
        next: while (i.hasNext()) {
            final FsController ec = i.next();
            final FsMountPoint emp = ec.getModel().getMountPoint();
            final FsManager fm = new FsFilteringManager(emp, manager);
            for (final FsController fc : fm) {
                final FsMountPoint fmp = fc.getModel().getMountPoint();
                if (mp.equals(fmp) || synced.contains(fmp)) {
                    if (emp.equals(fmp) || synced.contains(emp)) i.remove();
                    continue next;
                }
            }
            i.remove(); // even if subsequent umount fails
            fm.sync(FsSyncOptions.SYNC);
        }
    }

    /**
     * Registers the archive file system of the given controller as the most
     * recently used (MRU).
     *
     * @param controller the controller for the most recently used file system.
     */
    void accessed(final FsController controller) {
        if (controller.getModel().isMounted()) synced.add(controller);
    }

    @Override
    public void sync(final BitField<FsSyncOption> options)
    throws FsSyncException {
        evicted.clear();
        try {
            manager.sync(options);
        } finally {
            synced.mount(manager);
        }
    }

    @SuppressWarnings("serial")
    private static final class MountedControllerMap
    extends LinkedHashMap<FsMountPoint, FsController> {
        volatile int max =
                PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_DEFAULT_VALUE;
        final Collection<FsController> evicted;
        
        MountedControllerMap(final Collection<FsController> evicted) {
            super(INITIAL_CAPACITY, 0.75f, true);
            assert null != evicted;
            this.evicted = evicted;
        }

        @Override
        public boolean removeEldestEntry(
                final Map.Entry<FsMountPoint, FsController> entry) {
            final boolean evict = size() > max;
            if (evict) {
                final FsController c = entry.getValue();
                final boolean added = evicted.add(c);
                assert added;
            }
            return evict;
        }
    } // MountedControllerMap

    private static final class SyncedControllerSet {
        final Map<FsMountPoint, FsController> mounted;
        final ReentrantReadWriteLock.ReadLock readLock;
        final ReentrantReadWriteLock.WriteLock writeLock;

        SyncedControllerSet(final Map<FsMountPoint, FsController> mounted) {
            final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
            readLock = lock.readLock();
            writeLock = lock.writeLock();
            assert null != mounted;
            this.mounted = mounted;
        }

        boolean contains(FsMountPoint key) {
            readLock.lock();
            try {
                return mounted.containsKey(key);
            } finally {
                readLock.unlock();
            }
        }

        void add(final FsController controller) {
            final FsMountPoint mp = controller.getModel().getMountPoint();
            writeLock.lock();
            try {
                mounted.put(mp, controller);
            } finally {
                writeLock.unlock();
            }
        }

        int mount(final FsManager manager) {
            writeLock.lock();
            try {
                mounted.clear();
                for (final FsController c : manager)
                    if (c.getModel().isMounted())
                        mounted.put(c.getModel().getMountPoint(), c);
                return mounted.size();
            } finally {
                writeLock.unlock();
            }
        }
    } // SyncedControllerSet
}

File: truevfs-ext/truevfs-ext-pacemanager/src/main/scala/net/java/truevfs/ext/pacemanager/PaceManager.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.pacemanager

import java.{util => ju}
import java.util.concurrent._
import java.util.concurrent.locks._
import javax.annotation.concurrent._
import net.java.truecommons.shed._
import net.java.truevfs.comp.jmx._
import net.java.truevfs.kernel.spec._
import net.java.truevfs.kernel.spec.sl._
import PaceManager._

/**
 * The pace manager.
 *
 * @author Christian Schlichtherle
 */
@ThreadSafe
private final class PaceManager(mediator: PaceMediator, manager: FsManager)
extends JmxManager[PaceMediator](mediator, manager) {

  private[this] val evicted = new ConcurrentLinkedQueue[FsController]
  private[this] val mounted = new MountedControllerMap(evicted)
  private[this] val accessed = new AccessedControllerSet(mounted)

  def maximumFileSystemsMounted = mounted.max

  def maximumFileSystemsMounted_=(max: Int) {
    if (max < PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_MINIMUM_VALUE)
      throw new IllegalArgumentException
    mounted.max = max
  }

  protected override def newView = new PaceManagerView(this)

  /**
   * If the number of mounted archive files exceeds `maximumFileSystemsMounted`,
   * then this method `sync`s the least recently used (LRU) archive files which
   * exceed this value.
   *
   * @param  controller the controller for the file system to retain mounted
   *                    for subsequent access.
   */
  def retain(controller: FsController) {
    val it = evicted.iterator
    if (!it.hasNext) return
    val manager = FsManagerLocator.SINGLETON.get
    val mp = controller.getModel.getMountPoint
    while (it.hasNext) {
      val ec = it.next
      val emp = ec.getModel.getMountPoint
      val fm = new FsFilteringManager(emp, manager)
      def sync(): Boolean = {
        import collection.JavaConversions._
        for (fc <- fm) {
          val fmp = fc.getModel.getMountPoint
          if (mp == fmp || (accessed contains fmp)) {
            if (emp == fmp || (accessed contains emp)) it remove ()
            return false
          }
        }
        true
      }
      if (sync()) {
        it remove () // even if subsequent sync fails!
        fm sync FsSyncOptions.SYNC
      }
    }
  }

  /**
   * Registers the archive file system of the given controller as the most
   * recently used (MRU).
   *
   * @param controller the controller for the most recently used file system.
   */
  def accessed(controller: FsController) {
    if (controller.getModel.isMounted) accessed add controller
  }

  override def sync(options: BitField[FsSyncOption]) {
    evicted clear ()
    try {
      manager sync options
    } finally {
      accessed mount manager
    }
  }
}

private object PaceManager {
  private val initialCapacity = HashMaps.initialCapacity(
    PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_DEFAULT_VALUE + 1)

  private final class MountedControllerMap(evicted: ju.Collection[FsController])
  extends ju.LinkedHashMap[FsMountPoint, FsController](initialCapacity, 0.75f, true) {

    override def removeEldestEntry(entry: ju.Map.Entry[FsMountPoint, FsController]) =
      if (size > max) evicted.add(entry.getValue) else false

    @volatile
    var max = PaceManagerMXBean.MAXIMUM_FILE_SYSTEMS_MOUNTED_DEFAULT_VALUE
  } // MountedControllerMap

  private final class AccessedControllerSet(
    mounted: ju.Map[FsMountPoint, FsController],
    lock: ReentrantReadWriteLock = new ReentrantReadWriteLock
  ) {

    assert (null ne mounted)

    private[this] val readLock = lock.readLock
    private[this] val writeLock = lock.writeLock

    private def locked[V](lock: Lock)(operation: => V) = {
      lock lock ()
      try {
        operation
      } finally {
        lock unlock ()
      }
    }

    def contains(key: FsMountPoint) = locked(readLock)(mounted containsKey key)

    def add(controller: FsController) {
      val mp = controller.getModel.getMountPoint
      locked(writeLock)(mounted.put(mp, controller))
    }

    def mount(manager: FsManager) = {
      locked(writeLock) {
        mounted clear ()
        import scala.collection.JavaConversions._
        for (controller <- manager; model = controller.getModel)
          if (model.isMounted) mounted.put(model.getMountPoint, controller)
        mounted.size
      }
    }
  } // AccessedControllerSet
}
